common/network-common/src/main/java/org/apache/spark/network/sasl/aes/SparkAesCipher.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.sasl.aes;
+
+import java.io.IOException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Properties;
+import javax.crypto.Cipher;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+import javax.crypto.spec.IvParameterSpec;
+import javax.security.sasl.SaslException;
+
+import org.apache.commons.crypto.cipher.CryptoCipher;
+import org.apache.commons.crypto.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * AES cipher for encryption and decryption.
+ */
+public class SparkAesCipher {
```

Comment:
The key negotiation part is fine. That doesn't need to change.

But I'm strongly against your argument that because `SaslEncryption` exists you shouldn't change it. That's an internal Spark class, you have all the right to change it. If you change it to work with ByteBuffers, the AES path automatically becomes more efficient, while there's no gain or loss for the old SASL-based path. Yes it's more code churn, but for a good cause.

Similarly, all your arguments about needing to calculate an HMAC and keep a sequence number are based on the internal JVM implementation of DIGEST-MD5 encryption. That's irrelevant to what you're trying to achieve here. Meaning that code is not necessary.

Basically, being blunt, what I'm saying is that I'm opposed to your current approach and I will -1 your patch the way it is.